### ISSUE TITLE
Delete array of keys with prefix

### ISSUE THREAD

**Issue by artcups** [phpredis/phpredis#1117]
We have a redis cluster that serves multiple application caches and we prefix the keys with the system name with the setOption method. And in each system we can have several different caches hence the prefix variable.

The problem is when you want to delete all keys with a prefix.
The getKeys method returns an array with all the keys, but when we try to delete them it does´nt.

The code bellow is a simplified version of our setup that should find all the keys belonging to the system with a specific prefix.
```
function clearAppCacheByPrefix($prefix){
   $redis = new Redis();
   $redis->connect(REDIS_URL, REDIS_PORT);
   $redis->setOption(Redis::OPT_PREFIX, APP_REDIS_PREFIX); //APP_REDIS_PREFIX = QB:
   $redis->keys($prefix.'*'); //$prefix = exercise
   $redis->delete($keys);
   $redis->close();
}
```

The delete method returns an array with each key with the app prefix and the searched prefix eg.
```
0 = QB:exercise-3071df4c-1973-487a-b862-96883861c59e
1 = QB:exercise-23dd33sd-2313-123d-4ffd-dscq2ce45ggf
```

When i try to delete the keys in the array it doesent work with the keys found with the keys method. If i remove the QB: (APP_PREFIX) from the string the key is removed. 

It seems that the del function is also adding the prefix to the key but how am i then supposed to remove all keys to a specific prefix or pattern?

**Comment by michael-grunder** [comment]
Hey,

It's an interesting question.  As you're seeing, phpredis is going to prefix the pattern you send with `keys` and also prefix the key name sent to `del`.  

It isn't going to do string manipulation on the keys being returned from `keys` in order to remove the prefix, as it treats data coming back from `keys` as raw.  This is something that's come up a few times over the years.

There are a variety of ways you could get around this.  The simplest that comes to mind would be something like this:

```php
$keys = $r->keys($prefix.'*');

$oldprefix = $r->getOption(Redis::OPT_PREFIX);
$r->setOption(Redis::OPT_PREFIX, NULL);
$r->delete($keys);
$r->setOption(Redis::OPT_PREFIX, $oldprefix);
``` 

If you prefer not toggling the prefix setting, it should be pretty simple to scrub the values returned by `keys` as you know Redis::OPT_PREFIX will be at the start of the string, and how long it is.  Should be as simple as doing a substr on the returned values.

This is something that's come up before, but I'm not sure if it's functionality we would want to add to the library itself but i'm not totally against the idea.

Cheers,
Mike

**Comment by artcups** [comment]
Yes the fastest solution was, as you mentioned, substr on the keys. But it seems lika tedious solution to the problem :)

When i am using the prefix option I am not interested in the prefix on the keys hence the prefix option. 

How about a rawKeys method or a del function that takes a pattern?

BTW the lib is amazing, super easy to integrate and use.

**Comment by michael-grunder** [comment]
At this point we don't have plans to massage the prefix away from commands that return key names.

When you need to do that we suggest doing it in userland php code.
